Makefile

```
@@ -1,162 +1,217 @@
 
 .PHONY: run-storybook
-run-storybook: build-storybook  ## Build & Run storybook.
-	cd frontend && npm run storybook
+run-storybook: build-storybook  ## Build & Run storybook
+	@cd frontend && npm run storybook
+
+
+# TEST (Implicit builds)
 
-# TEST
 
 .PHONY: test
-test: test-frontend test-backend test-mdbook test-e2e  ## Test everything.
-	pre-commit
+test: test-frontend test-backend test-mdbook test-e2e  ## Test everything
+	$(MAKE) pre-commit
 
 .PHONY: test-e2e
-test-e2e: ## End-to-End tests. Needs install-e2e, backend and frontend running.
-	cd e2e && ./e2e.sh
+test-e2e:  ## End-to-End tests. Needs install-e2e, backend and frontend running
+	@cd e2e && ./e2e.sh
 
 .PHONY: test-frontend
-test-frontend:  ## Test Frontend.
-	cd frontend && npm install && npm run format:check && npm run lint && npm run test
+test-frontend:  build-frontend  ## Build & Test Frontend
+	@cd frontend && npm install && npm run format:check && npm run lint && npm run test
 
 .PHONY: test-backend
-test-backend:  ## Test Backend.
-	cd backend && make test
+test-backend:  build-backend  ## Build & Test Backend
+	@make test -C ./backend
 
 .PHONY: test-mdbook
-test-mdbook:  ## Test Mdbook.
-	mdbook test
+test-mdbook:  build-mdbook  ## Build & Test Mdbook
+	@mdbook test
+
+# Test all makefile targets except the ones with "run-".
+.PHONY: test-makefile
+test-makefile: $(MAKEFILE_LIST)
+	@for target in `grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | awk 'BEGIN {FS = ":.*?## "}; !/^run-/ { printf "%s\n", $$1 }'`; do \
+		echo "$(RED)Testing target: $$target$(RESET)"; \
+		$(MAKE) $$target || (echo "$(RED)Test for target: $$target failed$(RESET)" && exit 1); \
+	done
+	@echo "$(RED)All non 'run-' targets passed the test$(RESET)"
+
 
-.PHONY: test-storybook
-test-storybook:
-	@echo "\033[0;31m Not supported \033[0m" && exit 1
+# BUILD (Implicit installs)
 
-# BUILD
 
 .PHONY: build
-build: generate-api-types build-frontend build-backend build-storybook build-mdbook  # Build everything.
+build: generate-api-types build-frontend build-backend build-storybook build-mdbook  # Build everything
 
 .PHONY: build-frontend
-build-frontend: generate-api-types  ## Build frontend.
-	cd frontend && npm install && npm run generate-api-types && npm run build
+build-frontend: install-frontend generate-api-types  ## Build Frontend
+	@cd frontend npm run generate-api-types && npm run build
 
 .PHONY: build-backend
-build-backend: generate-api-types  ## Build backend.
-	cd backend && make build
+build-backend: install-backend generate-api-types  ## Build Backend
+	@make build -C ./backend
 
 .PHONY: build-mdbook
-build-mdbook:  ## Build mdbook.
-	mdbook build
+build-mdbook: install-backend  ## Build Mdbook
+	@mdbook build
 
 .PHONY: build-storybook
-build-storybook: install generate-api-types  ## Build storybook.
-	cd frontend && npm install && npm run build-storybook
+build-storybook: generate-api-types  ## Build Storybook
+	@cd frontend && npm install && npm run build-storybook
```

Here is a bug, to make the storybook complete, also npm run doc is needed before:

```suggestion
	@cd frontend && npm install && npm run doc && npm run build-storybook
```